### ISSUE TITLE
Multi-line functions not indented correctly.

### ISSUE THREAD

**Issue by mbush92** [ahernguo/urscript-extension#4]
We have an issue where if we use a function and have the parameters on multiple lines the function is not indented at all.  This causes readability issues.  

We are getting code that looks like this on save

```
someFunc(
a = accel,
v= speed,
useForce = True
)
```

when we want it to be formatted with indents for readability like this

```
someFunc(
  a = accel,
  v= speed,
  useForce = True
)
```

I tried updating needIncreaseIndent and needDecreaseIndent regex in formattingEditProviders.ts to include `|(\()` and `|(/))` respectively and that works for formatting correctly when you are typing but then the auto-format on save does not correct any indentation.  If I remove those additional regex groups then the auto-format on save works again but we don't get the code formatted properly.  Any help solving this would be greatly appreciated.

**Comment by mbush92** [comment]
I have found one solution that seems to work by changing the REGEX, I changed the increase indent regex to `\b(def|thread|while|for|if|elif|else).*:|\b\w+\((?!.*\))` and the decrease indent regex to `\b(?<!\w)end(?!\w)|(elif.*:)|(else:)|(?<!(\(.*))\)`. Not sure if there is another way that you would like to see this approached.  I had to add the lookahead and lookbehind regex to the `end` keyword in the decrease indent method to guard against issues where if that word was used in a variable or method name it would cause the line to remove the indentation.

**Comment by ahernguo** [comment]
I tried to fix without regex at [0d3c96b](https://github.com/ahernguo/urscript-extension/commit/0d3c96b21d769b0eb9324c5f0cee4022f845a8e9)
(It is too late to see your solution 😫)

I made `formatBracket` returns a `boolean` that brackets is closed or not. It will add more indent when it returns `true`.

`needDecreaseIndent` will judge a single '(' to remove indent.

Maybe we can compare two solution and keep one of best.

**Comment by ahernguo** [comment]
I updated [35741b5](https://github.com/ahernguo/urscript-extension/commit/35741b56bccbd2075a8b9eda0682df8a6c3e5a5f)

`formatBracket` now return the status of brackets. It shows add or remove indent more easily

**Comment by mbush92** [comment]
I tested the lastest code in master and it indents as we would expect.  Thanks for taking a look at that.  I did find one issue around the keyword `end` when it is used in variable or method names causes the indentation to be removed.  I have issued a PR for that.

**Comment by mbush92** [comment]
Found a case today while reformatting all of our code base that isn't indenting correctly, not sure why this is adding an indent level

```
def getParameter(name, requiredType):
  log_debug(str_cat("Name needed: ", name), str_cat(", Value received: ", parameter))
  log_trace(str_cat(str_cat(str_cat("Parameter `", name), "`, Expected type: `"), requiredType),   str_cat("`, Received: ", parameter))
    if (to_str(parameter) == name):
      return getDefaultType(requiredType)
    end
    return parameter
  end

```

**Comment by ahernguo** [comment]
I found the problem cause by `StringPattern` failed.

I'm update the temporary solution at [7a12a14](https://github.com/ahernguo/urscript-extension/commit/7a12a147fce64f535b91553043db6a93ea5721c8)
(fix miss last part at [1316709](https://github.com/ahernguo/urscript-extension/commit/13167098db81bfc7792f08f5a45bc1fa5d89523b))
(Update brackets regex at [9130202](https://github.com/ahernguo/urscript-extension/commit/91302022585b8b6a9d3e1a30cd7c425614afced1))

This solution is not perfect.
It may use regex to split directly. But I didn't get right regex to do this 😅

**Comment by mbush92** [comment]
Just merged master into our Fork and that last issue we added seems to be addressed.  We did find some other issues yesterday which I am working on a PR for, they are around formatting signs and stuff like that.

**Comment by ahernguo** [comment]
> Just merged master into our Fork and that last issue we added seems to be addressed. We did find some other issues yesterday which I am working on a PR for, they are around formatting signs and stuff like that.

Hi @mbush92 ,

I rewrote `formattingEditProvider.ts`. Now work more correctly 😄

The biggest change is ordering of `SignPatterns`. For example, check `==` before `=`.
And now check the line range where edited to avoid 'over range' error

1. It will not format `,`, `( )` and `[ ]` that in the string
```
origin → somefunc("apple,banana", count)
 
old version → somefunc("apple, banana", count)
new version → somefunc("apple,banana", count)
```

2. `if`, `elif`, `while` and others will be format
```
origin → if(abc==True)or(ijk==False)

old version → if(abc == True)or(ijk == False)
new version → if (abc == True) or (ijk == False)
```

3. Minus value
```
origin → somefunc(-d2r(value))

old version → somefunc( - d2r(value))
new version → somefunc(-d2r(value))
```

You may try my latest `master`. Hope it have solved your problem. 😀

**Comment by mbush92** [comment]
I saw the commit to master on Friday and definitiely works much better, thank you for the quick responses.  I did find a couple of new issues on Friday around the `-` sign. I posted them against the closed issue about adding extra spaces but I will post them here:  `-4.4E-4` becomes `- 4.4E - 4` and also if you have more than one negated number in a list it adds spaces on the subsequent ones so `[ 0, -0.1, -0.1 ]` becomes `[ 0, -0.1, - 0.1 ]`, there may also be some issues with a `-` following a `[` bracket, I noticed in the formatter that you do not have that character in the regex and the scientific number above does follow the `p[` portion of the pose as it is the X component.

Other than that our code compiles and runs, we did not see any other changes that we would not have expected.
